directorywatcher/directorywatcher_test.go

```diff
@@ -0,0 +1,147 @@
+package directorywatcher_test
+
+import (
+	"context"
+	"encoding/json"
+	"io"
+	"os"
+	"path/filepath"
+	"testing"
+	"time"
+
+	"github.com/reddit/baseplate.go/directorywatcher"
+	"github.com/reddit/baseplate.go/filewatcher"
+	"github.com/reddit/baseplate.go/internal/limitopen"
+	"github.com/reddit/baseplate.go/log"
+)
+
+func TestDirectoryWatcher(t *testing.T) {
+	dir := t.TempDir()
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*30)
+	defer cancel()
+	data, err := directorywatcher.New(
+		ctx,
+		directorywatcher.Config{
+			Path: dir,
+			Parser: func(f io.Reader) (interface{}, error) {
+				reader := f.(limitopen.ReadCloser)
+				var data map[string]interface{}
+				folder := make(map[string]interface{})
+				err := json.NewDecoder(f).Decode(&data)
+				if err != nil {
+					t.Fatal(err)
+				}
+
+				folder[reader.Path] = data
+				return folder, err
+			},
+
+			// a function that knows how to add a files data from the interface
+			Adder: func(d interface{}, file interface{}) (interface{}, error) {
+				if d == nil {
+					d = make(map[string]interface{})
+				}
+				folder := d.(map[string]interface{})
+
+				for key, value := range file.(map[string]interface{}) {
+					folder[key] = value
+				}
+				return folder, nil
+
+			},
+
+			// a function that can clean up the data based on path name
+			Remover: func(d interface{}, path string) (interface{}, error) {
+				folder := d.(map[string]interface{})
+				delete(folder, path)
+				return folder, nil
+			},
+
+			Logger: log.TestWrapper(t),
+		},
+	)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer data.Stop()
+
+	path1 := filepath.Join(dir, "foo")
+	if _, err = os.Create(path1); err != nil {
+		t.Fatal(err)
+	}
+	if err = os.WriteFile(path1, []byte("{\"number\":17}"), 0644); err != nil {
+		t.Fatal(err)
+	}
+
+	time.Sleep(1 * time.Second) // give the watcher time to parse
```

Comment:
Create a synchronous way to detect that it has parsed, rather than using a sleep.  It can be unexported if you want.

For example, a channel that, if non-nil, has a boolean pushed onto it whenever parsing completes.  Now you can have a `select` on that channel along with a timeout, so you can say `parse failed to execute within 1 second` if it times out but don't need to wait the full second if it happens fast.